lms/src/lms/private.rs

```diff
@@ -0,0 +1,352 @@
+
+    fn try_from(pk: &'a [u8]) -> Result<Self, Self::Error> {
+        if pk.len() < 4 {
+            return Err(LmsDeserializeError::NoAlgorithm);
+        }
+
+        let (alg, pk) = pk.split_at(4);
+        let expected = Mode::M + ID_LEN + 8;
+
+        // will never panic because alg is a 4 byte slice
+        if u32::from_be_bytes(alg.try_into().unwrap()) != Mode::TYPECODE {
+            return Err(LmsDeserializeError::WrongAlgorithm);
+        }
+
+        match pk.len().cmp(&expected) {
+            Ordering::Less => Err(LmsDeserializeError::TooShort),
+            Ordering::Greater => Err(LmsDeserializeError::TooLong),
+            Ordering::Equal => {
+                // pk is now guaranteed to be of the form otstype || q || id || seed
+                let (otstype, qk) = pk.split_at(ID_LEN);
+                let (q, idseed) = qk.split_at(4);
+                let (id, seed) = idseed.split_at(ID_LEN);
+
+                // check the OTS type
+                if u32::from_be_bytes(otstype.try_into().unwrap()) != Mode::OtsMode::TYPECODE {
+                    return Err(LmsDeserializeError::WrongAlgorithm);
+                }
+
+                let mut key = Self {
+                    q: u32::from_be_bytes(q.try_into().expect("ok")),
+                    id: id.try_into().expect("ok"),
+                    seed: GenericArray::clone_from_slice(seed),
+                    auth_tree: GenericArray::default(),
+                };
+                key.gen_pk_tree();
+                Ok(key)
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::SigningKey;
+    use crate::lms::modes::{LmsSha256M32H10, LmsSha256M32H5};
+    use crate::ots::modes::{LmsOtsSha256N32W4, LmsOtsSha256N32W8};
+    use hex_literal::hex;
+    use signature::{RandomizedSignerMut, SignatureEncoding};
+
+    #[test]
+    fn test_pk_tree_kat1() {
+        let seed = hex!("558b8966c48ae9cb898b423c83443aae014a72f1b1ab5cc85cf1d892903b5439");
+        let id = hex!("d08fabd4a2091ff0a8cb4ed834e74534");
+        let expected_k = hex!("32a58885cd9ba0431235466bff9651c6c92124404d45fa53cf161c28f1ad5a8e");
+
+        let lms_priv = SigningKey::<LmsSha256M32H10<LmsOtsSha256N32W4>>::new_from_seed(id, seed);
+        let lms_pub = lms_priv.public();
+        assert_eq!(lms_pub.k(), expected_k);
+        assert_eq!(lms_pub.id(), &id);
+    }
+
+    #[test]
+    fn test_pk_tree_kat2() {
+        let seed = hex!("a1c4696e2608035a886100d05cd99945eb3370731884a8235e2fb3d4d71f2547");
+        let id = hex!("215f83b7ccb9acbcd08db97b0d04dc2b");
+        let expected_k = hex!("a1cd035833e0e90059603f26e07ad2aad152338e7a5e5984bcd5f7bb4eba40b7");
+
+        let lms_priv = SigningKey::<LmsSha256M32H5<LmsOtsSha256N32W8>>::new_from_seed(id, seed);
+        let lms_pub = lms_priv.public();
+        assert_eq!(lms_pub.k(), expected_k);
+        assert_eq!(lms_pub.id(), &id);
+    }
+
+    #[test]
+    fn test_kat_2() {
+        let expected_signature = [
```

Comment:
Can you provide some information on the provenance of these KATs?

It appears to be [RFC8554 Appendix F](https://datatracker.ietf.org/doc/html/rfc8554#appendix-F) but I'm noticing `test_pk_tree_kat1` appears to be "Test Case 2".

Reply:
@tjade273